forte/train_preprocessor.py

```diff
@@ -0,0 +1,313 @@
+#  Copyright 2020 The Forte Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+from typing import Optional, Dict, Type, Any, Union, Iterator
+
+import torch
+from texar.torch.data import DataIterator, Batch
+from torch import device
+
+from forte.common.configuration import Config
+from forte.data.converter.converter import Converter
+from forte.data.data_pack_dataset import DataPackDataSource, \
+    DataPackDataset
+from forte.data.extractor.extractor import BaseExtractor
+from forte.data.extractor.predictor import Predictor
+from forte.data.extractor.unpadder import BaseUnpadder, SameLengthUnpadder
+from forte.data.ontology.core import Entry
+from forte.data.ontology.core import EntryType
+from forte.data.readers.base_reader import PackReader
+from forte.data.types import DATA_OUTPUT
+
+logger = logging.getLogger(__name__)
+
+
+class TrainPreprocessor:
+    """
+    `TrainPreprocessor` provides the functionality of doing pre-processing work
+    including building vocabulary, extracting the features, batching and
+    padding. It will provide methods to return the iterator over batch of
+    padded tensors.
+
+    `TrainPreprocessor` will maintain a Config that stores all the configurable
+    parameters for various components.
+
+    Particularly, the `dataset` config is exactly the same as what
+    texar.torch.data.DatasetBase::default_config will return.
+
+    `TrainPreprocessor` will also accept a user request. Internally it will
+    parse this user request and store the parsed result as a dictionary
+    into `feature_resource`.
+    An example `feature_resource` is:
+    feature_resource: {
+        "scope": ft.onto.Sentence
+        "schemes": {
+            "text_tag": {
+                "extractor":  Extractor,
+                "converter": Converter,
```

Comment:
I still think we should put it into `config`. Here are the reasons:

* `feature_resource` is actually the global context that will be passed to different class/instances (i.e. datapack_dataset, predictor). It would be convenient to put `Converter` here.

* Even though currently there is only one `Converter`, in the future users/developers probably want to extend it to have different conversion logic.

So IMO we shoud keep this. Any drawback I missed?

Reply:
Since the name `Converter`, `Extractor` is very generic, I thought you are going to have one only. But I do see different types of `Extractor` like `LinkExtractor`, do we have an actual class named `Extractor?

Reply:
No. There is one called `BaseExtractor`

Reply:
So the example here actually says we should put a concrete extractor here? In that case, we can use a real extractor to avoid confusion.

Reply:
Yes. Both `extractor` and `converter` here is the concrete instance. I will update the doc. So I am still bit confused. Can I keep put this `Converter` in this `feature_resource`?

Reply:
We can talk about this directly in the meeting.